Clean up XContentBuilder

tlrx opened the pull request:
This PR cleans most of the methods of XContentBuilder so that:

- ~~methods that actually write to the underlying `XContentGenerator` are now private and all prefixed with a "write" (like `writeNumber(long)`)~~
- Jackson's convenience methods are used instead of our custom ones (ie `field(String,long) `now uses Jackson's `writeNumberField(String, long)` instead of calling `writeField(String)` then `writeNumber(long)`)
- null checks are added for all field names and values
- methods are grouped by type in the class source
- methods have the same parameters names
- duplicated methods like `field(String, String...)` and `array(String, String...`) are removed in favor of the latter
- varargs methods now have the "array" name to reflect that it builds arrays
- unused methods like `field(String,BigDecimal)` are removed
- all methods now follow the execution path field(String,?) -> field(String) then value(?), and value(?) -> writeSomething() method. Methods to build arrays also follow the same execution path.
- ~~use optimized Jackson v2.8 methods to write int and long arrays~~

It also adds basic tests for these method so that Jackson updates will be more comfortable.

nik9000 commented:
Neat!

tlrx commented:
@nik9000 this is ready for review, do you want to have a look?

nik9000 commented on core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java:
```diff
@@ -117,658 +148,904 @@ public boolean isPrettyPrint() {
+            writer.write(this, value);
+        } else if (value instanceof Path) {
+            //Path implements Iterable<Path> and causes endless recursion and a StackOverFlow if treated as an Iterable here
+            value((Path) value);
```
Maybe for another commit - should we move some/all of these into the `WRITERS` map?

nik9000 replied:
I guess we're doing `instanceof` not `getClass()`, but still, it sounds like maybe we should in some cases.

tlrx replied:
The `Writer` is retrieve from the `WRITERS` map using the `getClass()` key so only concrete classes can be referenced there. We could add abstract classes or interfaces to the map too, but then we'd have to iterate over map entries to find the writer. Not sure this is better?

nik9000 replied:
Yeah. I'm not sure either. I like that concrete classes are nice and quick. Lets drop this point for this PR at least.

nik9000 commented on core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java:
```diff
@@ -117,658 +148,904 @@ public boolean isPrettyPrint() {
+     *
+     * Use {@link org.elasticsearch.common.xcontent.XContentParser#binaryValue()} to read the value back
+     */
+    public XContentBuilder value(BytesRef value) throws IOException {
```
Should this be `binaryValue` so there isn't a `value(BytesRef)` method? Calling this one `value` makes it look like it is the "normal" way to write `BytesRef` but I think actually `utf8Value` is more useful?

tlrx replied:
Good point - this will also be more coherent with the `XContentParser.binaryValue()` method

nik9000 commented on core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java:
```diff
@@ -117,658 +148,904 @@ public boolean isPrettyPrint() {
+        } else if (value.hasString()) {
+            return value(value.string());
+        } else {
+            // TODO: TextBytesOptimization we can use a buffer here to convert it? maybe add a
```
Is it worth making an issue for this TODO so we don't forget about it?

tlrx replied:
I created #20381

nik9000 commented on core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java:
```diff
@@ -117,658 +148,904 @@ public boolean isPrettyPrint() {
+    //
+    ////////////////////////////////////////////////////////////////////////////
+
+    private void writeStartObject() throws IOException {
```
Should these be final?

nik9000 replied:
While you are cleaning up maybe we should put these into their own class that `XContentBuilder` extends? That way all of the wrapper methods are all in one spot.

tlrx replied:
> While you are cleaning up maybe we should put these into their own class that XContentBuilder extends? That way all of the wrapper methods are all in one spot.

I'd prefer not: these methods already delegate to XContentGenerator which delegates to JsonGenerator etc, so I prefer to keep this simple.

Since all the methods have been grouped by types and correctly delegate to each other, I think there's no more need to group the writing method separate. We could just merge them with their calling method and have:

```
public XContentBuilder value(boolean value) throws IOException {
        generator.writeBoolean(value);
        return this;
}
```

instead of 

```
public XContentBuilder value(boolean value) throws IOException {
        writeBoolean(value);
        return this;
}

private void writeBoolean(boolean value) throws IOException {
        generator.writeBoolean(value);
        return this;
}
```

We'll just have to take care when adding new methods that we don't fall back in the previous "let's duplicate everything" style.

nik9000 commented on core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java:
```diff
@@ -786,156 +1063,28 @@ public BytesReference bytes() {
-            writeValue(v);
-        }
-        generator.writeEndArray();
+    public static XContentBuilder builder(XContent xContent) throws IOException {
```
Can you stick these two methods at the top and javadoc them?

tlrx replied:
Sure

nik9000 commented on core/src/main/java/org/elasticsearch/common/xcontent/XContentGenerator.java:
```diff
@@ -20,14 +20,13 @@
 import java.io.InputStream;
 
-/**
- *
```
🎉

nik9000 commented on core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java:
```diff
@@ -19,23 +19,665 @@
+        expectObjectException(() -> builder().endObject().string());
+
+        expectValueException(() -> builder().startObject("foo").endObject());
+        expectNonNullFieldException(() -> builder().startObject().startObject(null).endObject().endObject());
```
Does this need the `.endObject().endObject()` bit?

tlrx replied:
No, let's remove it

nik9000 commented on core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java:
```diff
@@ -182,4 +824,71 @@ protected void doTestBigInteger(JsonGenerator generator, ByteArrayOutputStream o
+    }
+
+    public static org.hamcrest.Matcher<String> equalToJSon(String json) {
+        return org.hamcrest.core.IsEqual.equalTo(json.replace("'", "\""));
```
`Matchers.equalTo` or `equalTo` with a static import is more normal for this sort of thing I think.

nik9000 commented on core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java:
```diff
@@ -182,4 +824,71 @@ protected void doTestBigInteger(JsonGenerator generator, ByteArrayOutputStream o
+    }
+
+    private static void expectUnclosedException(ThrowingRunnable runnable) {
+        IllegalStateException e = expectThrows(IllegalStateException.class, runnable);
```
Awesome.

nik9000 commented on core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java:
```diff
@@ -19,23 +19,665 @@
+
+        assertResult(() -> builder().startObject().endObject(), equalTo("{}"));
+        assertResult(() -> builder().startObject().startObject("foo").endObject().endObject(), equalToJSon("{'foo':{}}"));
+        assertResult(() -> builder().startObject().startObject("foo").startObject("bar").endObject().endObject().endObject(),
```
When it gets long like this can you indent it like
```
assertResult(() -> builder()
    .startObject()
        .startObject("foo")
            .startObject("bar")
            .endObject()
        .endObject()
    .endObject(),
```

I know we `assertThat` has the matcher second, but maybe we should put the closure second for this? I think it is nice when the closure is second because it makes the code formatting prettier.

tlrx replied:
We're already using a custom `assertResult` method so yes, let's do things easier and have expected string first, then closure.

nik9000 commented on core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java:
```diff
@@ -19,23 +19,665 @@
+        expectFieldException(() -> builder().startObject().startArray().endArray().endObject());
+        expectNonNullFieldException(() -> builder().startObject().startArray(null).endArray().endObject());
+
+        assertResult(() -> builder().startObject().startArray("foo").endArray().endObject(), equalToJSon("{'foo':[]}"));
```
Maybe add one with values in the array?

nik9000 commented on core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java:
```diff
@@ -19,23 +19,665 @@
+    }
+
+    public void testNullValue() throws IOException {
+        assertResult(() -> builder().startObject().field("foo").nullValue().endObject(), equalToJSon("{'foo':null}"));
```
Can you make it `equalToJson`? I like the rules [here](https://google.github.io/styleguide/javaguide.html#s5.3-camel-case) because it makes it easy to tell where the words start and end.

tlrx replied:
It's a mistake, I used it dozens of time without even notice the `S`. Thanks!

nik9000 commented on core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java:
```diff
@@ -19,23 +19,665 @@
+        builder.endObject();
+
+        XContentParser parser = xcontentType().xContent().createParser(builder.bytes());
+        assertThat(parser.nextToken(), is(Token.START_OBJECT));
```
`assertEquals` and `assertSame` for these? I like to avoid hamcrest if it doesn't buy anything beyond "normal" junit asserts.

nik9000 commented on core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java:
```diff
@@ -19,23 +19,665 @@
+                equalToJSon("{'t3':'" + formatter.print(t3) + "'}"));
+        assertResult(() -> builder().startObject().field("t3").value(t3, formatter).endObject(),
+                equalToJSon("{'t3':'" + formatter.print(t3) + "'}"));
+
```
Is it worth adding a random test here? I'd love to have tests for far future and far past dates and that'd cover it.

nik9000 replied:
Maybe add tests for other `ReadableInstant` like `Instant`?

tlrx replied:
Sure, I added both

nik9000 commented on core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java:
```diff
@@ -19,23 +19,665 @@
+        assertResult(() -> builder().startObject().field("d3").value(d3, formatter).endObject(),
+                equalToJSon("{'d3':'" + formatter.print(d3.getTime()) + "'}"));
+
+        expectThrows(IllegalArgumentException.class, () -> builder().startObject().field("d3", d3, null).endObject());
```
Can you assert something about the message? As it stands someone could break something to throw a **different** `IllegalArgumentException` and we'd not notice.

tlrx replied:
Sure. I've been lazy here.

nik9000 commented on core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java:
```diff
@@ -19,23 +19,665 @@
+        assertResult(() -> builder().startObject().field("d3", d3, formatter).endObject(),
+                equalToJSon("{'d3':'" + formatter.print(d3.getTime()) + "'}"));
+        assertResult(() -> builder().startObject().field("d3").value(d3, formatter).endObject(),
+                equalToJSon("{'d3':'" + formatter.print(d3.getTime()) + "'}"));
```
You use this version of this method so much, maybe you should make a version of `assertResult` that takes a string?

nik9000 commented on core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java:
```diff
@@ -19,23 +19,665 @@
+            assertResult(() -> builder().startObject().array("objects", o.getValue()).endObject(), matcher);
+        }
+
+        assertResult(() -> builder().startObject().array("objects", null, null, null).endObject(),
```
I think this is a duplicate of the above?

tlrx replied:
It is, I remove it

nik9000 commented on core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java:
```diff
@@ -19,23 +19,665 @@
+        final Calendar c1 = new DateTime(2010, 1, 1, 0, 0, DateTimeZone.UTC).toCalendar(Locale.ROOT);
+        object.put("{'object':'2010-01-01T00:00:00.000Z'}", c1);
+
+        final ToXContent x1 = (builder, params) -> builder.startObject().field("f1", "v1").field("f2", 2).array("f3", 3, 4, 5).endObject();
```
I love that you can do this in Java 8 now. Much, much nicer than it used to be.

tlrx replied:
+1

nik9000 commented:
Awesome! I left a bunch of small stuff, mostly arounds testing.

nik9000 commented:
Also, I'm happy about all the new testing!

tlrx commented:
@nik9000 Thanks for your review! I updated the PR according to your comments. I think I addressed most of them; for the remaining ones I replied as a comment.

I also moved back from using Jackson's `writeArray(long[])` and  `writeArray(int[])` methods. I noticed an issue using  `writeArray(double[])` and I'm wondering if it's a Jackson bug. Since these methods have been added to Jackson recently and are only optimized for Smile & CBOR I think we can wait a bit before using them here.

nik9000 commented:
LGTM. Do you want to cherry pick this to the 5.0 and 5.x branches? I expect it is fairly safe and it'll be nice to have this code consistent across those two branches.

tlrx commented:
> Do you want to cherry pick this to the 5.0 and 5.x branches?

Yes, it makes sense.

tlrx commented:
Backport on 5.x in 0e5a5e804b624ce8df4069cfb74e86ddf853cc69 and on 5.0 in 69b3c611581a0be3256ecc758de3bab9cc1e8d41
